### PR TITLE
MFA-related Language Changes & Move MFA logic to OSS 

### PR DESCRIPTION
**Pull request by rudream**
## Purpose

Changed the way we call MFA devices after discussing with @klizhentas and @kimlisa .

Moved MFA logic to OSS for the upcoming MFA device management feature and also to reduce the scope of the PR in `webapps.e`.

## Implementation

The back-end MFA-related endpoints have been moved to the OSS cfg, and I created a new `MfaService` for MFA-related logic, this service will also be used by the account dashboard device management.

We previously had inconsistent and sometimes confusing naming when it came to MFA devices. We now call them "Two-Factor Devices".

- "TOTP" is now "Authenticator App".
- "U2F" is now "Hardware Key".

The flows for resetting `totp` and `u2f` have been combined into one, there is no longer a "Lost U2F Key?"/"Lost Two-Factor Token?", but a single "Lost Two-Factor Device?".

### Login

login

#### U2F

image

#### TOTP

image

### CODE REVIEW

File: packages/teleport/src/services/mfa/makeMfaDevice.ts
Comment (kimlisa) [inline packages/teleport/src/services/mfa/makeMfaDevice.ts:21]: hm, is this conversion required?

File: packages/teleport/src/services/mfa/types.ts
Comment (kimlisa) [inline packages/teleport/src/services/mfa/types.ts:3]: i'm not really seeing the benefit of typing this, do we use this else where other than in `makeMfaDevice`? If not, i think it's better to just contain it inside the makeXXX?

File: packages/teleport/src/config.ts
Comment (kimlisa) [inline packages/teleport/src/config.ts:120]:
```suggestion
    mfaDevicePath: '/v1/webapi/mfadevice',
```

if it was something like `/v1/webapi/mfadevice/delete` then `mfaDeleteDevicePath` would be better

File: packages/teleport/src/config.ts
Comment (kimlisa) [inline packages/teleport/src/config.ts:119]:
```suggestion
    mfaDeviceListPath: '/v1/webapi/mfadevice/:tokenId',
```
to keep it consistent elsewhere, i know there is like a few that deviates, but I go with majority

File: packages/teleport/src/components/MfaDeviceList/MfaDeviceList.tsx
Comment (kimlisa) [inline packages/teleport/src/components/MfaDeviceList/MfaDeviceList.tsx:127]: i'd remove this until we need it, also long comments should go on top of the field, side comments are usually very short

File: packages/teleport/src/components/MfaDeviceList/MfaDeviceList.tsx
Comment (rudream) [reply packages/teleport/src/components/MfaDeviceList/MfaDeviceList.tsx:45]: It definitely looks better with the radius, what I will do is only have the radius if `isRecoveryFlow`.

File: packages/teleport/src/services/mfa/makeMfaDevice.ts
Comment (rudream) [reply packages/teleport/src/services/mfa/makeMfaDevice.ts:21]: Yes. The `addedAt` and `lastUsed` returned from the backend are formatted strings eg. `"2021-08-18T18:39:23.522007Z"`. Converting them to unix timestamps makes sorting possible.

File: packages/teleport/src/services/mfa/types.ts
Comment (rudream) [reply packages/teleport/src/services/mfa/types.ts:3]: We do use it in a couple places in `webapps.e`

File: packages/shared/components/FormInvite/TwoFaInfo.tsx
Comment (alex-kovoy) [inline packages/shared/components/FormInvite/TwoFaInfo.tsx:51]: there is `caps` property that can be used instead of textTransform

File: packages/shared/components/FormLogin/FormLogin.tsx
Comment (alex-kovoy) [inline packages/shared/components/FormLogin/FormLogin.tsx:291]: keep the strings as it's way more descriptive than a boolean flag. -> `password | 2fa`

File: packages/teleport/src/components/MfaDeviceList/MfaDeviceList.tsx
Comment (alex-kovoy) [inline packages/teleport/src/components/MfaDeviceList/MfaDeviceList.tsx:95]: Please use `px` units as this is what we use everywhere

File: packages/teleport/src/components/MfaDeviceList/MfaDeviceList.tsx
Comment (alex-kovoy) [inline packages/teleport/src/components/MfaDeviceList/MfaDeviceList.tsx:104]: It's better to avoid any dynamically resolved property name (like this `Text` suffix). It's hard to read, it's hard to `type` (Typescript), and easy to break.

File: packages/teleport/src/components/MfaDeviceList/MfaDeviceList.tsx
Comment (alex-kovoy) [inline packages/teleport/src/components/MfaDeviceList/MfaDeviceList.tsx:109]: `DeleteDeviceBtn` is not a button but a `Cell` -> `DeleteCell`

File: packages/teleport/src/services/mfa/makeMfaDevice.ts
Comment (alex-kovoy) [reply packages/teleport/src/services/mfa/makeMfaDevice.ts:21]: @rudream no need to use `unix` nor `moment` here, just create `Date` objects as follows
 
```
  registeredDate: new Date(addedAt),
  lastUsedDate: new Date(lastUsed),
```

File: packages/teleport/src/services/mfa/mfa.ts
Comment (alex-kovoy) [inline packages/teleport/src/services/mfa/mfa.ts:14]: deleteMfdDevice -> `removeDevice` (the service name already has `Mfa` prefix)
fetchMfaDevices -> `fetchDevices`

File: packages/teleport/src/services/mfa/types.ts
Comment (alex-kovoy) [reply packages/teleport/src/services/mfa/types.ts:3]: this is a strange name for a type, please remove any spaces

File: packages/teleport/src/services/mfa/types.ts
Comment (alex-kovoy) [inline packages/teleport/src/services/mfa/types.ts:11]: DeleteMfaDeviceRequest -> `RemoveDeviceRequest`

File: packages/teleport/src/services/mfa/types.ts
Comment (rudream) [reply packages/teleport/src/services/mfa/types.ts:3]: @alex-kovoy this is just for the text on the `type` column in the table, I will rename `type` to `typeLabel` to make it more clear

File: packages/teleport/src/services/mfa/types.ts
Comment (alex-kovoy) [inline packages/teleport/src/services/mfa/types.ts:4]: I am a bit confused with this `typeText`. Why does it need to be `typed` ? If this is a description then `description: string` should be enough. 

```
  id: string;
  name: string;
  type:  'TOTP' | 'U2F' ...,
  description: string;
```

File: packages/teleport/src/services/mfa/mfa.ts
Comment (alex-kovoy) [inline packages/teleport/src/services/mfa/mfa.ts:13]: No need for `{ ...data} `

```ts
removeDevice(req: RemoveDeviceRequest) {
    return api.delete(cfg.api.mfaDevicePath, req);
  }
```

File: packages/teleport/src/services/mfa/types.ts
Comment (rudream) [reply packages/teleport/src/services/mfa/types.ts:4]: @alex-kovoy In the table we display either "Hardware Key" or "Authenticator App"

image

The alternative is using `type: 'TOTP' | 'U2f'` and then creating something like a `DeviceTypeCell` that checks the `type` and then for example renders `"Hardware Key"` if `type === 'U2F`, but we won't use the strings `'TOTP'` or `'U2F'` anywhere else, so I am just doing that check on the service layer and returning the device object with the formatted text already there.

File: packages/teleport/src/components/MfaDeviceList/MfaDeviceList.tsx
Comment (alex-kovoy) [inline packages/teleport/src/components/MfaDeviceList/MfaDeviceList.tsx:25]: since it uses `useMemo` what happens to it when a new devices added?

File: packages/teleport/src/components/MfaDeviceList/MfaDeviceList.tsx
Comment (alex-kovoy) [inline packages/teleport/src/components/MfaDeviceList/MfaDeviceList.tsx:140]: is it really needed for this shared component to be aware of recovery flow and account dashboard?

File: packages/teleport/src/components/MfaDeviceList/MfaDeviceList.tsx
Comment (alex-kovoy) [inline packages/teleport/src/components/MfaDeviceList/MfaDeviceList.tsx:47]: instead of `isRecoveryFlow`, you can propagate the props to the table

```tsx
const {
  devices = [],
  onDelete,
 ...styles
}: Props)

return (
<StyledTable data={data} ...styles>
```    

this way, a consumer of this component can apply custom styles

```tsx
<MfaDeviceList style={...} />
```

File: packages/teleport/src/components/MfaDeviceList/MfaDeviceList.tsx
Comment (alex-kovoy) [inline packages/teleport/src/components/MfaDeviceList/MfaDeviceList.tsx:126]: lets show the delete button by default

Comment (rudream) [reply packages/teleport/src/components/MfaDeviceList/MfaDeviceList.tsx:126]: @alex-kovoy This was a UX suggestion from Sasha, he wants us to hide the remove button for the most recent device

File: packages/teleport/src/components/MfaDeviceList/MfaDeviceList.tsx
Comment (rudream) [reply packages/teleport/src/components/MfaDeviceList/MfaDeviceList.tsx:140]: revised

File: packages/teleport/src/components/MfaDeviceList/MfaDeviceList.tsx
Comment (rudream) [reply packages/teleport/src/components/MfaDeviceList/MfaDeviceList.tsx:25]: I have changed the implementation, `mostRecentDevice` is now going to be provided by the parent `Devices` component in Recovery Flow. 

It is not possible for new devices to be added on this step since this `mostRecentDevice` is only used when the device list is in the recovery flow (not the dashboard), it is set when the list of devices is retrieved from the backend.

File: packages/teleport/src/services/mfa/utils.ts
Comment (kimlisa) [inline packages/teleport/src/services/mfa/utils.ts:10]: shouldn't `auth2faType === 'off'` return an empty array?

**Review by kimlisa — APPROVED** [review]
lgtm, please wait for @alex-kovoy approval

File: packages/teleport/src/services/mfa/utils.ts
Comment (rudream) [reply packages/teleport/src/services/mfa/utils.ts:10]: When `auth2faType === 'off'` there won't be a dropdown shown, but behind the scenes this allows us to use `if(mfaOption.value === 'optional)'` both when the 'None' option is selected and when `auth2faType === 'off'` rather than needing an additional check for the latter

**Review by alex-kovoy — APPROVED** [review]

**Pull request merged**
